### ISSUE TITLE
Importing second time not working

### ISSUE THREAD

**Issue by MattiasNilssonBeckhoff** [iokiwi/jsd-bulk-customer-upload#14]
Hi @iokiwi 

Imported something around 1500 organizations and 3000 customers (or something like that) and was going to import the rest of the customers today. However the bulk import gets stuck. 

Directly after starting the import get this:
`2020-03-11 13:02:41,622 - INFO - bulk_customer_import.organization - Retrieving organisations`

but nothing happens. When running debug mode these two lines keep repeating:
```
2020-03-11 13:02:48,486 - DEBUG - urllib3.connectionpool - http://bas-jira01:8080 "GET /rest/servicedeskapi/organization HTTP/1.1" 200 None
2020-03-11 13:02:48,489 - DEBUG - urllib3.connectionpool - Starting new HTTP connection (1): bas-jira01:8080
```

After aborting (ctrl-c):
```
Traceback (most recent call last):
  File "C:\Users\MattiasN\AppData\Local\Programs\Python\Python38\lib\runpy.py", line 193, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Users\MattiasN\AppData\Local\Programs\Python\Python38\lib\runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "C:\TEMP\bulk_customer_import\__main__.py", line 2, in <module>
    foo.main()
  File "C:\TEMP\bulk_customer_import\bulk_customer_import.py", line 61, in main
    organizations = client.organization.list()
  File "C:\TEMP\bulk_customer_import\organization.py", line 12, in list
    return self.client.get_paginated_resource(
  File "C:\TEMP\bulk_customer_import\client.py", line 77, in get_paginated_resource
    response = self.get(url, **kwargs).json()
  File "C:\TEMP\bulk_customer_import\client.py", line 63, in get
    response = self.request("GET", url, **kwargs)
  File "C:\TEMP\bulk_customer_import\client.py", line 46, in request
    return requests.request(
  File "C:\Users\MattiasN\AppData\Local\Programs\Python\Python38\lib\site-packages\requests\api.py", line 61, in request
    return session.request(method=method, url=url, **kwargs)
  File "C:\Users\MattiasN\AppData\Local\Programs\Python\Python38\lib\site-packages\requests\sessions.py", line 530, in request
    resp = self.send(prep, **send_kwargs)
  File "C:\Users\MattiasN\AppData\Local\Programs\Python\Python38\lib\site-packages\requests\sessions.py", line 643, in send
    r = adapter.send(request, **kwargs)
  File "C:\Users\MattiasN\AppData\Local\Programs\Python\Python38\lib\site-packages\requests\adapters.py", line 439, in send
    resp = conn.urlopen(
  File "C:\Users\MattiasN\AppData\Local\Programs\Python\Python38\lib\site-packages\urllib3\connectionpool.py", line 665, in urlopen
    httplib_response = self._make_request(
  File "C:\Users\MattiasN\AppData\Local\Programs\Python\Python38\lib\site-packages\urllib3\connectionpool.py", line 421, in _make_request
    six.raise_from(e, None)
  File "<string>", line 3, in raise_from
  File "C:\Users\MattiasN\AppData\Local\Programs\Python\Python38\lib\site-packages\urllib3\connectionpool.py", line 416, in _make_request
    httplib_response = conn.getresponse()
  File "C:\Users\MattiasN\AppData\Local\Programs\Python\Python38\lib\http\client.py", line 1322, in getresponse
    response.begin()
  File "C:\Users\MattiasN\AppData\Local\Programs\Python\Python38\lib\http\client.py", line 303, in begin
    version, status, reason = self._read_status()
  File "C:\Users\MattiasN\AppData\Local\Programs\Python\Python38\lib\http\client.py", line 264, in _read_status
    line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
  File "C:\Users\MattiasN\AppData\Local\Programs\Python\Python38\lib\socket.py", line 669, in readinto
    return self._sock.recv_into(b)
KeyboardInterrupt
```

**Comment by iokiwi** [comment]
Hey @MattiasNilssonBeckhoff  remind me, it's Jira Service Desk Server `4.7.1` right?

**TLDR; I think its working, just taking a while to get 1500 organisations due to pagination - I'll add some more logging so there's some visibility over its progress**

When you say "stuck" how long are you letting it sit for before cancelling?

I suspect it's just taking a really long time to download all of the organisations - which is the first thing it does. (Its not a great way to do it but I think it was the only way at the time - I should revisit this though)

It's 'stuck' inside `get_paginated_resource` which traverses the API's pagination mechanism to get all the records for a resource/

```
 File "C:\TEMP\bulk_customer_import\client.py", line 77, in get_paginated_resource
    response = self.get(url, **kwargs).json()
```

There are 1500 organisations so depending on the pagination size it could have to make around 150 requests to get them all.

The fact that you keep seeing this line means it is making a request for each page of the results.

```
2020-03-11 13:02:48,486 - DEBUG - urllib3.connectionpool - http://bas-jira01:8080 "GET /rest/servicedeskapi/organization HTTP/1.1" 200 None
```

If I add some logging we will at least be able to make sure its not stuck requesting the same page over and over.

I will find some time in the next few hours to add logging for you.

But I think - if you leave it for long enough...it will get there.

Regards,
Simon

**Comment by iokiwi** [comment]
@MattiasNilssonBeckhoff I pushed a branch with logging for the part that gets stuck here.

https://github.com/iokiwi/jsd-bulk-customer-upload/tree/new/paginated-resource-logging

Can you pull the branch and run this version with`-l debug`. I expect the output to look something like this.

```
2020-03-11 13:02:48,486 - INFO - bulk_customer_upload.client - Retrieving resource(s) starting from 0"
2020-03-11 13:02:48,486 - DEBUG - urllib3.connectionpool - http://bas-jira01:8080 "GET /rest/servicedeskapi/organization HTTP/1.1" 200 None
2020-03-11 13:02:48,489 - DEBUG - urllib3.connectionpool - Starting new HTTP connection (1): bas-jira01:8080
2020-03-11 13:02:48,486 - INFO - bulk_customer_upload.client - Retrieving next 10 resource(s)  10 - 20"
2020-03-11 13:02:48,486 - DEBUG - urllib3.connectionpool - http://bas-jira01:8080 "GET /rest/servicedeskapi/organization HTTP/1.1" 200 None
2020-03-11 13:02:48,489 - DEBUG - urllib3.connectionpool - Starting new HTTP connection (1): bas-jira01:8080
```

I will try push a version of the script which checks if organisations exist in a more intelligent manner than retrieving them all up front.

**Comment by MattiasNilssonBeckhoff** [comment]
@iokiwi Tried it to be running yesterday for a little more than one hour. Tried it today with your new code, aborted it just before it retrieved resources from 1 million. 
`2020-03-12 12:02:00,090 - INFO - bulk_customer_import.client - Retrieving next 50 resource(s) 974050 - 974100

2020-03-12 12:02:00,160 - INFO - bulk_customer_import.client - Retrieving next 50 resource(s) 974100 - 974150
2020-03-12 12:02:00,224 - INFO - bulk_customer_import.client - Retrieving next 50 resource(s) 974150 - 974200
2020-03-12 12:02:00,285 - INFO - bulk_customer_import.client - Retrieving next 50 resource(s) 974200 - 974250`

This is the output from debug:

`2020-03-12 12:02:33,588 - DEBUG - bulk_customer_import.bulk_customer_import - Client {'platform': 'server', 'verify': None, 'base_url': 'http://bas-jira01:8080/'}

2020-03-12 12:02:33,591 - INFO - bulk_customer_import.organization - Retrieving organisations
2020-03-12 12:02:33,591 - INFO - bulk_customer_import.client - Retrieving all resources from paginated endpoint.this may take a while
2020-03-12 12:02:33,591 - INFO - bulk_customer_import.client - Retrieving resource(s) starting from 0
2020-03-12 12:02:33,591 - DEBUG - bulk_customer_import.client - request params: {'start': 0}
2020-03-12 12:02:33,592 - DEBUG - urllib3.connectionpool - Starting new HTTP connection (1): bas-jira01:8080
2020-03-12 12:02:33,676 - DEBUG - urllib3.connectionpool - http://bas-jira01:8080 "GET /rest/servicedeskapi/organization HTTP/1.1" 200 None
2020-03-12 12:02:33,677 - INFO - bulk_customer_import.client - Retrieving next 50 resource(s) 50 - 100
2020-03-12 12:02:33,678 - DEBUG - bulk_customer_import.client - request params: {'start': 50}
2020-03-12 12:02:33,679 - DEBUG - urllib3.connectionpool - Starting new HTTP connection (1): bas-jira01:8080
2020-03-12 12:02:33,743 - DEBUG - urllib3.connectionpool - http://bas-jira01:8080 "GET /rest/servicedeskapi/organization HTTP/1.1" 200 None
2`

**Comment by iokiwi** [comment]
Ok my appologies @MattiasNilssonBeckhoff , there's clearly something broken with pagination. I will dig deeper over the weekend.

Its incrementing the `start` parameter but maybe something has changed in the response. It goes untill the API responds with `isLastPage` in the response data but maybe that condition is no longer being met...which means the response from that API has changed.

I will spin up my test server again tonight and debug. 

Sorry for the issues and thanks for your patience.

**Comment by iokiwi** [comment]
@MattiasNilssonBeckhoff was able to reproduce locally by creating a test `example.csv` with more than 50 lines.

Solution was a small fix to one line https://github.com/iokiwi/jsd-bulk-customer-upload/blob/master/bulk_customer_import/client.py#L92

**Fix has been merged to master. https://github.com/iokiwi/jsd-bulk-customer-upload/pull/15**

Let me know if I can help you out with anything else :)

**Comment by MattiasNilssonBeckhoff** [comment]
Worked perfect now, awesome! :)

**Comment by iokiwi** [comment]
Thanks for the confirmation @MattiasNilssonBeckhoff :)
